Lots of NCRC HowTo TLC (#22446)
idaholab/moose: modules/doc/content/ncrc/ncrc_local_binary.md

  -h, --help            show this help message and exit
```

## Common Usage

##### List all available NCRC applications

```bash
 $> ncrc list

# Use 'ncrc search name-of-application' to list more detail

> # NCRC applications available:

Reviewer: Did you mean for this to be the highest level of heading?

Author: That happens to be a comment being generated by the `ncrc` script:

```pre
\```bash
 $> ncrc list

# Use 'ncrc search name-of-application' to list more detail
# NCRC applications available:

pronghorn
bison
griffin
sabertooth
relap7
\```
```

Author: I've been re-doing the indentation going on. I don't much care for the itemized list justification going on, for only half the page...

Reviewer: Oh sorry, I completely missed that it was part of the output..... You can disregard this one. I liked how you moved away from so many itemized lists as well.

pronghorn
bison
griffin
sabertooth
relap7
```

##### Lists all available versions of a specific NCRC application

```bash
$> ncrc search griffin

Loading channels: done
# Name                       Version           Build  Channel
ncrc-griffin              2021_07_29         build_0  ncrc-applications
ncrc-griffin              2021_08_27         build_0  ncrc-applications
ncrc-griffin              2021_08_31         build_0  ncrc-applications
ncrc-griffin              2021_09_03         build_0  ncrc-applications
ncrc-griffin              2021_09_09         build_0  ncrc-applications
ncrc-griffin              2021_09_15         build_0  ncrc-applications
ncrc-griffin              2021_09_25         build_0  ncrc-applications
ncrc-griffin              2021_10_02         build_0  ncrc-applications
```

##### Install an NCRC application

> Installing an NCRC application requires authentication. If you have sufficient privileges the package installation will commence:

Reviewer: ```suggestion
Installing an NCRC application requires authentication. If you have sufficient privileges, the package installation will commence:
```

```bash
$> conda activate base
$ (base) >
$ (base) > ncrc install bison

Username: johndoe
PIN+TOKEN:
Installing bison...
```

##### Install a specific NCRC application version

> Don't want to use the latest? For whatever the reason, you can install a specific version of an available NCRC application in the following way:

Reviewer: ```suggestion
Don't want to use the latest version? For whatever the reason, you can install a specific version of an available NCRC application in the following way:
```

```bash
$> conda activate base
$ (base) >
$ (base) > ncrc install griffin=2021_09_15

Username: johndoe
PIN+TOKEN:
Installing griffin=2021_09_15...
```

##### Update an NCRC application

To update your NCRC application to the latest available, you first must activate the environment it is in, and then perform the update:

```bash
$> conda activate bison
$ (bison) > ncrc update bison

Username: johndoe
PIN+TOKEN:
```

> ##### Uninstalling NCRC application

Reviewer: ```suggestion
##### Uninstalling an NCRC application
```

> The NCRC script being a wrapper tool, is unable to perform such a function. The user must deactivate the environment

Reviewer: ```suggestion
The NCRC script, being a wrapper tool, is unable to perform such a function. The user must deactivate the environment
```

and remove that environment using the appropriate conda commands instead. The following is an error scenario instructing
the user how to achieve the task:

```bash
$> ncrc remove bison

Due to the way ncrc wraps itself into conda commands, it is best to
remove the environment in which the application is installed. Begin
by deactivating the application environment and then remove it:
  conda deactivate
  conda env remove -n bison
```

## Using the Application

Once your NCRC application is installed, you have access to use said binary and/or view the documentation.

!alert note

> Where applicable, you only need to activate the environment once. Each example below assume the environment has not yet been activated.

Reviewer: ```suggestion
Where applicable, you only need to activate the environment once. Each example below assumes the environment has not yet been activated.
```

triso_pebble.cylinder_2d_rz ............................................................................... OK
triso_pebble.cylinder_3d .................................................................................. OK
triso_pebble.sphere_2d_rz ................................................................................. OK
triso_pebble.sphere_3d .................................................................................... OK
triso_pebble.plate ........................................................................................ OK
--------------------------------------------------------------------------------------------------------------
Ran 1606 tests in 274.6 seconds. Average test time 0.7 seconds, maximum test time 20.7 seconds.
1606 passed, 22 skipped, 0 pending, 0 failed
```

##### Some Applications contain more than tests

Some applications have different inputs that can be copied and executed. Each application can use whatever naming convention
they wish, so listing all possibilities is impossible. However, any runnable input directory will have a file called `testroot`

> residing in it. The following *crazy* command will return those possibilities (using Bison as an example):

Reviewer: 🤪

Author: yeah... hahahaha any help here in making this... better described is appreciated!

Reviewer: Added another comment with a wording suggestion that you can take or leave. I appreciate that you are giving a resource for the user to find what those input categories actually are.

Author: Yeah, I then need to add what to actually do with these. Basically another example of how one runs tests.

> Some applications have different inputs that can be copied and executed. Each application can use whatever naming convention
> they wish, so listing all possibilities is impossible. However, any runnable input directory will have a file called `testroot`
> residing in it. The following *crazy* command will return those possibilities (using Bison as an example):

Reviewer: ```suggestion
Some applications have different inputs that can be copied and executed using the `copy-inputs` flag. Each application can use whatever naming convention
they wish for these input categories, so listing all possibilities is impossible. However, any runnable input directory will have a file called `testroot`
residing in it, in which the test categories are listed. The following piped command sequence will return those possibilities (using Bison as an example):
```

```bash
$> conda activate bison
$ (bison) > find $CONDA_PREFIX/$NCRC_APP/share/$NCRC_APP -maxdepth 2 -type f -name 'testroot' \
| awk 'BEGIN { FS = "/" } ; { print $(NF-1) }'
workshop
tools
tests
assessment
examples
benchmarks
```

##### Run a specific input file

Unquestionably the real reason you installed said NCRC application, you can run your own input files for solving in the following way:

```bash
$> conda activate bison
$ (bison) >
$ (bison) > cd ~/path/to/my/input/file
$ (bison) > bison-opt -i my_input.i
# Using MPI
$ (bison) > mpiexec -n 8 bison-opt -i my_input.i
```

##### View Documentation

> Documentation exists as part of the binary installation process. You can access documentation by instructing your favorite web browser to open the index.html file.

Reviewer: ```suggestion
Documentation exists as part of the binary installation process. You can access documentation by instructing your favorite web browser to open the `index.html` file.
```

> We have made finding the location of this file via an environment variable, prefix with the NCRC application name:

Reviewer 2: ```suggestion
We have made finding the location of this file easier via an environment variable, prefixed with the NCRC application name:
```

Author: Its very much necessary in fact. Griffin for example my need/require a different set of libraries/versions of libMesh at the time the application was built.

Reviewer 2: ok let's say it explicitly somewhere then? 

unless I missed it

Author: Well, actually I am wrong in how I originally defended the requirement. But they still should be separated. The ncrc assumes the application name is also the environment name.

Author: Yeah I'll make mention of this in the first alert on the page. In my defense, the ncrc script itself will actually warn/error when the user tries to use it while not in the base environment (except during upgrades).

Reviewer: More content regarding separate environments for each app was added in subsequent commits. I think this is resolved.

> We have made finding the location of this file via an environment variable, prefix with the NCRC application name:

Reviewer: ```suggestion
We have made finding the location of this file via an environment variable, prefixed with the NCRC application name:
```
